articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+
+Checking for and cleaning up whitespace may be useful for your application, but shouldn't be relied upon as a secure way of avoiding Cypher injection.
+
+=== Unicode Encoding
+
+Another common exploit around input validation and sanitization is unicode encoding.
+Unicode encoding is where characters are encoded into their unicode equivalent.
+For example; the single quote character `'` can be encoded as `\u0027`.
+When sanitizing a string for the removal of escape quotation characters, it is important to also check for the unicode equivalent.
+The following query doesn't look like it escapes the string at first glance:
+
+`"Robby\u0027 MATCH (s:Student) DETACH DELETE s //"`
+
+But in reality, Cypher will resolve the unicode into a single quote and treat it as such in the compilation of the query.
+
+As mentioned earlier, malicious users may double encode their unicode input by additionally encoding the backslash, `\`, and `u`,
+which allows that to be resolved first in your client application code, and then resolved as the actual attack in the database.
+For example; `\u005C\u00750027` may be sent to the database as the input `\u0027`, thus successfully escaping the single quotation.
+
+When it comes to the validation of input such as usernames, it is often done to avoid the use of reserved keywords, such as admin.
+Unicode encoding can be used as another common bypass for this.
+For example the user input `\u0061\u0064\u006d\u0069\u006e` is the unicode for `admin`:
 
-Consider this query:
 
 [source,cypher]
 ----
-CALL apoc.cypher.doIt("CREATE (s:Student) SET s.name = '" + $studentName + "' RETURN true", {}) YIELD value
-RETURN value;
+CREATE (n {username: '\u0061\u0064\u006d\u0069\u006e'}) RETURN n.username
 ----
 
-Even though we passed `$studentName` as a parameter to the query, the parameter is being appended into the query string for execution by `apoc.cypher.doIt()`.
-The same vulnerability would exist if we were processing a LOAD CSV query like before, and appending into a query string for execution by APOC.
+.Results
+[opts="header"]
+|===
+|n.username
+|"Admin"
+|===
+
+=== String Concatenation
+
+Another method to bypass validation of particular keywords is using string concatenation in the injection.
+For example, validation that the user isn’t setting their username to admin could be bypassed with the injection:
+
+`"ad' + 'min'"`
+
+=== Second Order Injections
+
+A second order injection occurs when the input is successfully filtered and sanitized the first time it is used and is
+then stored in the database, when the application uses the value another time, the malicious code is then executed.
 
-Little Robby Labels would end up wiping out all our student data.
+For example; Little Robby Labels sets up an account with a username and password:
 
-== Pass parameters to APOC procs instead of appending into the query string
+`Username: LilRob' OR 1=1 WITH true AS hacked MATCH (a) DETACH DELETE ALL; –`
```

Comment:
```suggestion
`LilRob' OR 1=1 WITH true AS hacked MATCH (a) DETACH DELETE ALL; –`
```

Reply:
Inconsistent variables used between `a` and `ALL`, need to decide on only one for the entire example for consistency.

Reply:
Also a single dash is not a valid comment character, so the remaining query would error out. Maybe that would still let the attack succeed, but in general we should use the correct characters to comment out the remaining Cypher.

Reply:
oops started doing SQL haha clearly read too many articles :) Will fix up the naming inconsistencies!